### PR TITLE
[12.0][MIG] l10n_de_tax_statement

### PR DESCRIPTION
**Pull request by astirpe**
This is the porting to V12 of module `l10n_de_tax_statement` (German VAT Statement)

### CODE REVIEW

File: l10n_de_tax_statement/README.rst [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+.. contents::
+   :local:
+
+Installation
+============
+
+To install this module, you need to:
+
+#. Install module *l10n_de_tax_statement* version >= *11.0.1.0.0*.
```
Comment (Guenzn) [inline l10n_de_tax_statement/README.rst]: should be version >= 12.0.1.0.0

Comment (astirpe) [reply l10n_de_tax_statement/README.rst]: Done!

File: l10n_de_tax_statement/readme/CONFIGURE.rst [hunk 2]
```diff
@@ -0,0 +1,8 @@
...
+
+If the default Odoo German chart of accounts is installed (module l10n_de) you are able to select in the settings if you want to use the skr03 or skr04 chart variant. By installing and configuring your favored german account chart the tax tags from the module l10n_de are automatically present in the database. If this is the case, go to menu: Invoicing -> Configuration -> Accounting -> German Tax Tags, and check that the tags are correctly set. Click finally Apply to confirm the right tax - tax tags mappings.
+
+If a non-standard chart of accounts is installed, you have to manually create the tax tags and properly set them into the tax definition. If you create another german account chart (f.e. l10n_de_ikr) you can still depend on l10n_de module in order to benefit from the generic tax tags for germany. If you won't use l10n_de as a base module you have to configure at first your own tax tags. After that, go to go to menu: Invoicing -> Configuration -> Accounting -> German Tax Tags, and manually set the tags in the configuration form; click Apply to confirm (for more information about the installation and configuration of that module, check the README file).
```
Comment (Guenzn) [inline l10n_de_tax_statement/readme/CONFIGURE.rst:6]: I would hint here, that it is required to click Apply the first time!

Comment (astirpe) [reply l10n_de_tax_statement/readme/CONFIGURE.rst:6]: > click Apply to confirm

Should be enough IMO.

File: l10n_de_tax_statement/README.rst [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+If a non-standard chart of accounts is installed, you have to manually create the tax tags and properly set them into the tax definition. If you create another german account chart (f.e. l10n_de_ikr) you can still depend on l10n_de module in order to benefit from the generic tax tags for germany. If you won't use l10n_de as a base module you have to configure at first your own tax tags. After that, go to go to menu: Invoicing -> Configuration -> Accounting -> German Tax Tags, and manually set the tags in the configuration form; click Apply to confirm (for more information about the installation and configuration of that module, check the README file).
+
+The user must belong to the *Show Full Accounting Features* group, to be able to access the `Invoicing -> Configuration -> Accounting -> NL BTW Tags` menu.
```
Comment (Guenzn) [inline l10n_de_tax_statement/README.rst]: Should be "DE" for German localization?

Comment (astirpe) [reply l10n_de_tax_statement/README.rst]: Yes, but in this case the name of the menu is: `Ust-Voranmeldung`. Fixed!

Comment (astirpe) [reply l10n_de_tax_statement/README.rst]: :thinking: but in other places of the readme file there's written `German Tax Tags` for the same menu. I'm going to set `German Tax Tags` in the menu and update the readme accordingly.

Comment (Guenzn) [reply l10n_de_tax_statement/README.rst]: Yeah, menu names needs to be the same in readme as in odoo.  :+1:

Comment (astirpe) [reply l10n_de_tax_statement/README.rst]: Ok, should be correct now:
`Invoicing -> Configuration -> Accounting -> German Tax Tags`.

File: l10n_de_tax_statement/models/l10n_de_tax_statement_config.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
+# Copyright 2019 BIG-Consulting GmbH(<http://www.openbig.org>)
+# Copyright 2019 Onestein (<http://www.onestein.eu>)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import fields, models
+
+
+class VatStatementConfig(models.Model):
```
Comment (Guenzn) [inline l10n_de_tax_statement/models/l10n_de_tax_statement_config.py:8]: Modelname differs a little bit from class name sometimes....

**Review by Guenzn — COMMENTED** [review]
Good but sometimes class and model names are not exactly the same.
Did a code review not a functional

File: l10n_de_tax_statement/models/l10n_de_tax_statement_config.py (hunk 4 above)
Comment (astirpe) [reply l10n_de_tax_statement/models/l10n_de_tax_statement_config.py:8]: In the [OCA Guidelines](https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst) is not mentioned that the class name should be the same as the model name.

Changing the class name in UpperCamelCase would end up in "_L10nDeTaxStatementConfig_". Not so readable IMO. Is it ok to keep the class name as it is?

**Comment by astirpe** [comment]
Extracted from the document "_2019 Berlin OCA Days Spring Sprint_":


> The module l10n_de_tax_statement_zm is not exactly fitting the german report template of elsteronline.de. As we have cloned and adopted the module from netherlands the report (columns) does fit to netherlands standard. In the netherlands they have two columns to separate intra-community sales towards one partner separating the service sales in one column and the physical goods sales into the second column. 
> In germany we have just column for the invoices amount (for services + goods). Furthermore there is a column where we have to specify if this invoiced amount was a service or a goods invoicing. Here is the original form view:
> 
> https://www.bzst.de/DE/Steuern_International/USt_Kontrollverfahren_ZM_eCommerce/Zusammenfassende_Meldungen/Formulare/ZM_Formular_PDF.pdf?__blob=publicationFile . Here i want to point your interest to column “3” on side 2.

**Comment by pedrobaeza** [comment]
Something similar is being discussed in #44

**Comment by astirpe** [comment]
@pedrobaeza thanks!
Maybe splitting this PR in two separate PRs, one for each module, would help.

**Comment by astirpe** [comment]
Original PR is now split in two: the migration of `l10n_de_tax_statement_zm` (German VAT Statement Extension) is moved into PR #50

**Comment by haconsult** [comment]
Hi, 
it seems as if the numbers of fields do not match te statement form on elster-online. Please check e.g. field #53. It is "Umsatzsteuer" but in elster-online it is"Vorsteuerbeträge aus Rechnungen von anderen Unternehmern...". All following numbers do not match.

**Comment by astirpe** [comment]
@tv-openbig Would you to take a look at https://github.com/OCA/l10n-germany/pull/39#issuecomment-520252864?

**Comment by tv-openbig** [comment]
> Hi,
> it seems as if the numbers of fields do not match te statement form on elster-online. Please check e.g. field #53. It is "Umsatzsteuer" but in elster-online it is"Vorsteuerbeträge aus Rechnungen von anderen Unternehmern...". All following numbers do not match.

Yes indeed if you compare my screenshot you can see on left side the official "2018" version and on the right side the "2019" version. Current version in Odoo fits to 2018 edition of tax declaration if you regard the line numbering.

Bildschirmfoto 2019-08-12 um 14 54 08

However much more important in practise as the line numbering are the two digits code numbers formatted in bold letters in front of the base amount / tax amount values. Indeed there was obviously a removal of lines 49 /  51 (with their corresponding code numbers 52,53 & 78,79). In 2019 version they have merged the both lines with their corresponding tax codes under "old" line 52 (code number 84 / 85). 

At the moment i have no idea how we should handle those changes in the official tax 
declarations happening from year to year (deleted lines / deleted or merged tax cases) in our module. Maybe we would have to copy, split & modify the module "l10n_de_tax_statement" into "l10n_de_tax_statement_2018" and "l10n_de_tax_statement_2019", "l10n_de_tax_statement_2020" with separated menus, views.

Furthermore some taxes in the account charts defined to declare the lines 49 / 51 (2018) would be obsolete in 2019 but used in 2018 invoices. In addition the tax code naming in Odoo pointing to code numbers 84/85 in elster 2019 (old line 52 in 2018) needs to be changed as it represents in 2019 more / other tax cases as in 2018. Finally we would have to change l10n_de_skr03 and l10n_de_skr04 as well, minimum the tax namings. 

It seems the adoption of just the line numbers in the tax statement config wizards view and the tax reports view would be the easy part to "fix". However if someone wants to print the 2018 values some lines could be missing. @astirpe Do you have any idea ? How do you adopt those changes in netherland (l10n_nl_vat_statement) ?

**Comment by astirpe** [comment]
@tv-openbig 

For the changes regarding the line numbering and tax codes of version 2018 and version 2019, I could just allow the user to select the desired layout (2018, 2019, etc...) when creating/modifying the tax statement. The actual module can be reused, no need to create extra modules. This way, if someone needs to print an old version, it will be just re-printed as it was originally intended to be.

Technically I could just re-organize the actual python code of the module into multiple files, one for each version (eg.: `l10n_de_tax_statement_2018.py`, `l10n_de_tax_statement_2019.py`, etc...). This way the module will be easy to be extended in the future when a new version of text declaration will be published.

About merging or adding new lines, we should make sure that the calculations are done correctly, reflecting the changes made on the standard german localization modules (`l10n_de_skr03` and `l10n_de_skr04`). Do you already have the new german localization modules? Could you provide me with the code somehow?

**Comment by tv-openbig** [comment]
@astirpe 

I was not aware about the line changes. Personally i do the mapping to the lines in elsteronline.de web portal by checking the tax code which is represented in the report lines in brackets.  The specific §13b tax cases where the change of lines appeared should be relatively seldom, therefore i didn't stumbled over it yet and most people should not have any issues to enter the right values in elsteronline.de portal. Anyway I will check the required changes in l10n_de_tax_statement.py this week. I assume some modifications in between the lines 265 - 442 and finally in the "number of the beast" line 666. ;-) 

To my opinion we also need a modified tag_ids field in the german account charts.

As an example. The following tax needs a change in the account chart, as the tag_ids entry "Ust-VA KZ 52"  should be "Ust-VA KZ 84" starting from 01/01/2019, if that tax will be used in an invoice in 2019. I am afraid we will be able to adopt that change to all existing invoices in 2019, if we would apply such a change not exactly at the date 01/01/2019 by also taking the risk to get wrong tax tags for the case we will create or modify invoices in the prior year. More worse we would get inconsistent reports for prior years if the tax tags are not saved at the time of invoice creation as a field in account.move.line.

Bildschirmfoto 2019-08-12 um 17 56 08

Is it possible to adopt the potential issues i mentioned by modifying the l10n_de_tax_statement.py source code instead of changing the tag_ids ?

**Comment by astirpe** [comment]
> Is it possible to adopt the potential issues i mentioned by modifying the l10n_de_tax_statement.py source code instead of changing the tag_ids ?

Yes, in my opinion this is the way to go :+1:

**Comment by tv-openbig** [comment]
@astirpe Should we start here: https://github.com/OCA/l10n-germany/tree/11.0/l10n_de_tax_statement
I will make two copies from l10n_de_tax_statement.py:
l10n_de_tax_statement_2018.py (no change)
l10n_de_tax_statement_2019.py (changes according to official formular version for the year 2019).
Do you agree ?

**Comment by astirpe** [comment]
@tv-openbig you could start here: https://github.com/onesteinbv/l10n-germany/tree/12_mig_l10n_de_tax_statement
I already split the `l10n_de_tax_statement_2018.py` and the `l10n_de_tax_statement_2019.py`. You could just modify the `l10n_de_tax_statement_2019.py`.

In the meanwhile I'm thinking about how to handle the config form.

**Comment by tv-openbig** [comment]
Hi @astirpe ,

would you please read my comment written as a commit notice. I don't think that we could avoid changing the tag_ids configuration (which tag is assigned to which tax) and the configuration wizards. For example the following tax tag is obsolete as it is now merged under another tax_tag (for example all taxes formerly married with tag_52_base should now be married with tag_84_base). 

Furthermore as a logical consequence the configuration wizard needs adoptions too 
(as some lines in the wizards view are obsolete too).

Bildschirmfoto 2019-08-13 um 12 51 46

Did i missed something ?

File: l10n_de_tax_statement/models/l10n_de_tax_statement_2019.py [hunk 5]
```diff
@@ -282,39 +260,37 @@ def _finalize_lines_2019(lines):
...
+    lines['51']['tax'] = lines['51']['tax'] * 1
+    _51b = lines['51']['tax'] + _26b + _27b + _28b + _30b + _33b + _34b \
+           + _35b + _36b + _48b + _49b + _50b
+    # calculate line 60
+    lines['60']['tax'] = lines['60']['tax'] * 1
+    _60b = lines['60']['tax'] + _51b - _53b - _54b - _55b \
+           - _56b - _57b - _58b - _59b
+    # calculate line 64
+    lines['64']['tax'] = lines['64']['tax'] * 1
+    _64b = lines['64']['tax'] + _60b + _62b + _63b
```
Comment (astirpe) [inline l10n_de_tax_statement/models/l10n_de_tax_statement_2019.py:292]: @tv-openbig  here _62b and _63b are not defined, it seems the definition was deleted. Could you check?

File: l10n_de_tax_statement/models/l10n_de_tax_statement_2019.py [hunk 6]
```diff
@@ -138,99 +132,87 @@
...
     },
-    '66': {
-        'code': '66', 'tax': 0.0,
+    '64': {
+        'code': '64', 'tax': 0.0,
         'name': _('Umsatzsteuer-Vorauszahlung'),
     },
-    '67': {
-        'code': '67', 'tax': 0.0,
+    '65': {
```
Comment (astirpe) [inline l10n_de_tax_statement/models/l10n_de_tax_statement_2019.py]: Is line 65 to be mapped with tag 39? If this is the case, the mapping is also missing in old version 2018.

File: l10n_de_tax_statement/models/l10n_de_tax_statement_2019.py [hunk 7]
```diff
@@ -282,15 +282,65 @@ def _finalize_lines_2019(lines):
...
     _64b = lines['64']['tax'] + _60b + _62b + _63b
     # update lines 51, 60, 64
     lines['64'].update({'tax': _64b})
     lines['51'].update({'tax': _51b})
     lines['60'].update({'tax': _60b})
+
+
+def _get_tags_map_2019(config):
+    return {
+        config.tag_41_base.id: ('20', 'base'),
```
Comment (astirpe) [inline l10n_de_tax_statement/models/l10n_de_tax_statement_2019.py:301]: @tv-openbig could you double-check if the mapping tags-lines is correct?

**Comment by astirpe** [comment]
@tv-openbig configuration of obsolete tags should be kept for retro-compatibility, in case someone is printing old 2018 statements.
Merged tags can be managed by calculating their sum in method `_finalize_lines_2019()`.

With latest commit I simplified the configuration form: now it doesn't depend on lines anymore; it's just a simple mapping of tags.

File: l10n_de_tax_statement/models/l10n_de_tax_statement_2019.py [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+    lines['53']['tax'] = lines['53']['tax'] * -1
+    _53b = lines['53']['tax']
+    lines['54']['tax'] = lines['54']['tax'] * -1
+    _54b = lines['54']['tax']
+    lines['55']['tax'] = lines['55']['tax'] * -1
+    _55b = lines['55']['tax']
+    lines['56']['tax'] = lines['56']['tax'] * -1
+    _56b = lines['56']['tax']
+    lines['57']['tax'] = lines['57']['tax'] * -1
+    _57b = lines['59']['tax']
```
Comment (astirpe) [inline l10n_de_tax_statement/models/l10n_de_tax_statement_2019.py]: Should _57b be renamed to _59b?

File: l10n_de_tax_statement/models/l10n_de_tax_statement_2019.py [hunk 9]
```diff
@@ -0,0 +1,346 @@
...
+    lines['58']['tax'] = lines['58']['tax'] * -1
+    _58b = lines['60']['tax']
```
Comment (astirpe) [inline l10n_de_tax_statement/models/l10n_de_tax_statement_2019.py]: Should _58b be renamed to _60b?

File: l10n_de_tax_statement/models/l10n_de_tax_statement_2019.py [hunk 10]
```diff
@@ -0,0 +1,346 @@
...
+
+}
+
+
+def _finalize_lines_2019(lines):
+    _26b = lines['26']['tax']
+    _27b = lines['27']['tax']
+    _28b = lines['28']['tax']
+    _64b = lines['64']['tax']
+    _65b = lines['65']['tax']
```
Comment (astirpe) [inline l10n_de_tax_statement/models/l10n_de_tax_statement_2019.py]: _65b seems never used

File: l10n_de_tax_statement/models/l10n_de_tax_statement_2019.py (hunk 5 above)
Comment (tv-openbig) [reply l10n_de_tax_statement/models/l10n_de_tax_statement_2019.py:292]: Correct, in next commit i will change that.

File: l10n_de_tax_statement/models/l10n_de_tax_statement_2019.py (hunk 6 above)
Comment (tv-openbig) [reply l10n_de_tax_statement/models/l10n_de_tax_statement_2019.py]: This line is special and should be used only in December and the value should be entered manually. Therefore i assume in 2018 version it was line 67.

File: l10n_de_tax_statement/models/l10n_de_tax_statement_2019.py (hunk 8 above)
Comment (tv-openbig) [reply l10n_de_tax_statement/models/l10n_de_tax_statement_2019.py]: No that was not my intention. I have corrected those values.

File: l10n_de_tax_statement/models/l10n_de_tax_statement_2019.py (hunk 10 above)
Comment (tv-openbig) [reply l10n_de_tax_statement/models/l10n_de_tax_statement_2019.py]: That is the line i already explained above (to be manually entered)

File: l10n_de_tax_statement/models/l10n_de_tax_statement_2019.py (hunk 9 above)
Comment (tv-openbig) [reply l10n_de_tax_statement/models/l10n_de_tax_statement_2019.py]: Also an error from my side.

File: l10n_de_tax_statement/models/l10n_de_tax_statement_2019.py (hunk 7 above)
Comment (tv-openbig) [reply l10n_de_tax_statement/models/l10n_de_tax_statement_2019.py:301]: I have deleted some lines, as we have some deleted lines in the offical VAT formular (elsteronline.de portal). However i assume that we have to assign other tags to some taxes (removal of tax tags "68" -> now "60" // "52" and "78" -> now "84" // "53" and "79" -> now "85"). All taxes with those tags needs to be revised (however just from 01/01/2019 not before !!!). How should we address that ?

Comment (astirpe) [reply l10n_de_tax_statement/models/l10n_de_tax_statement_2019.py:301]: @tv-openbig the lines you just removed https://github.com/OCA/l10n-germany/pull/39/commits/ba91812cb91b84434d8ca9320865a7f2ebf11ffa were instead doing the job:

- tag [68](https://github.com/OCA/l10n-germany/pull/39/commits/ba91812cb91b84434d8ca9320865a7f2ebf11ffa#diff-83ae6e5803d61484e5e11e55c1d1e4a1L323), mapping line 39 (same as tag 60)
- tags [52](https://github.com/OCA/l10n-germany/pull/39/commits/ba91812cb91b84434d8ca9320865a7f2ebf11ffa#diff-83ae6e5803d61484e5e11e55c1d1e4a1L329) and [78](https://github.com/OCA/l10n-germany/pull/39/commits/ba91812cb91b84434d8ca9320865a7f2ebf11ffa#diff-83ae6e5803d61484e5e11e55c1d1e4a1L333), mapping line 50 (same as tag 84)
- tags [53](https://github.com/OCA/l10n-germany/pull/39/commits/ba91812cb91b84434d8ca9320865a7f2ebf11ffa#diff-83ae6e5803d61484e5e11e55c1d1e4a1L330) and [79](https://github.com/OCA/l10n-germany/pull/39/commits/ba91812cb91b84434d8ca9320865a7f2ebf11ffa#diff-83ae6e5803d61484e5e11e55c1d1e4a1L334), mapping line 50 (same as tag 85)

Could you check?

Comment (tv-openbig) [reply l10n_de_tax_statement/models/l10n_de_tax_statement_2019.py:301]: I have again added the deleted lines.

**Comment by astirpe** [comment]
@tv-openbig may I ask you your final review? After that, I will backport the latest developments to V11.

@haconsult Your issue should be fixed: this module now supports the old and the new statement format. Could you check?

**Review by tv-openbig — COMMENTED** [review]
Thanks @astirpe .I will make intense manual testing based on V11, after your backport.

**Comment by tv-openbig** [comment]
@astirpe May i merge your pull request ?

**Comment by astirpe** [comment]
Yes, I don't have other changes to make. Thanks!

**Comment by haconsult** [comment]
Works, thank you!

**Comment by haconsult** [comment]
I don't know if this is the right place, however there is a new issue.
When I created more than one Tax Report I can not go back to the tax report menu. It produces the following issue:

Fehler:
Odoo Server Error

Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/api.py", line 1039, in get
    value = self._data[key][field][record._ids[0]]
KeyError: 1

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/fields.py", line 981, in __get__
    value = record.env.cache.get(record, self)
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/api.py", line 1041, in get
    raise CacheMiss(record, field)
odoo.exceptions.CacheMiss: ('l10n.de.tax.statement(1,).tax_total', None)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/models.py", line 4718, in ensure_one
    _id, = self._ids
ValueError: too many values to unpack (expected 1)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/http.py", line 656, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/http.py", line 314, in _handle_exception
    raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/tools/pycompat.py", line 87, in reraise
    raise value
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/http.py", line 698, in dispatch
    result = self._call_function(**self.params)
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/http.py", line 346, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/service/model.py", line 97, in wrapper
    return f(dbname, *args, **kwargs)
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/http.py", line 339, in checked_call
    result = self.endpoint(*a, **kw)
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/http.py", line 941, in __call__
    return self.method(*args, **kw)
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/http.py", line 519, in response_wrap
    response = f(*args, **kw)
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/addons/web/controllers/main.py", line 904, in search_read
    return self.do_search_read(model, fields, offset, limit, domain, sort)
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/addons/web/controllers/main.py", line 926, in do_search_read
    offset=offset or 0, limit=limit or False, order=sort or False)
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/models.py", line 4589, in search_read
    result = records.read(fields)
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/models.py", line 2791, in read
    vals[name] = convert(record[name], record, use_name_get)
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/models.py", line 5117, in __getitem__
    return self._fields[key].__get__(self, type(self))
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/fields.py", line 985, in __get__
    self.determine_value(record)
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/fields.py", line 1098, in determine_value
    self.compute_value(recs)
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/fields.py", line 1052, in compute_value
    self._compute_value(records)
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/fields.py", line 1043, in _compute_value
    getattr(records, self.compute)()
  File "/usr/local/bin/custom-addons/l10n_de_tax_statement/models/l10n_de_tax_statement.py", line 370, in _compute_tax_total
    if self.version == '2019':
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/fields.py", line 979, in __get__
    record.ensure_one()
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/models.py", line 4721, in ensure_one
    raise ValueError("Expected singleton: %s" % self)
ValueError: Expected singleton: l10n.de.tax.statement(1, 2, 3)

**Comment by astirpe** [comment]
Thanks for reporting the error. I will fix it tomorrow.

**Comment by astirpe** [comment]
@haconsult fixed in #58 and #59

**Comment by haconsult** [comment]
Works great, thank you!

**Pull request merged**